caliban/hooks/util.py

```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/python
+#
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Functions to execute hooks
+"""
+
+from __future__ import absolute_import, division, print_function
+
+from typing import Any, Dict, List, NamedTuple, NewType, Optional, Union
+
+import importlib
+from absl import logging
+import json
+import sys
+
+import caliban.util.fs as ufs
+
+
+def perform_prebuild_hooks(caliban_config: Dict[str, Any]) -> Dict[str, str]:
+  """ Performs pre-build hooks (specified in the caliban_config) and:
+    1) If these run without error, accumulates their outputs into a
+       combined dictionary and returns it
+    2) If one of them errors, immediately stops and throws an Exception
+       displaying the error message for the user.
+
+  Prebuild hooks are python functions, which
+    1) Take no arguments, and
+    2) Output a dictionary with 'Succeeded' bool indicating success, and
+       a. 'Error' str in the case of failure
+       b. 'Data' Dict[str, str] in the case of success
+    3) Live in the 'hooks' module in the project's root directory
+  """
+  import sys
+  sys.path.append('.')
```

Comment:
I agree this is a little precarious. Again would be nice to hear @sritchie and @ajslone thoughts about the nicest way to structure this. 

Some options
- Have function take a path; have an optional hook folder path in the .calibanconfig.json; use os.os.getcwd from within the build and run calls if no path is given.
- Have a  required hook_folder path in  .calibanconfig.json. This could be all that's specified and we could move the specific hook specification to a hook_config.json within that directory.
- Have setup involve installing hooks in env.